docs/frontend-performance.md

```diff
@@ -0,0 +1,2984 @@
+  };
+}
+
+const selectAccountByAddress = createSelector(
+  (state, address) => state.metamask.internalAccounts.accounts.byAddress[address.toLowerCase()],
+  (state) => state.metamask.internalAccounts.accounts.byId,
+  (accountId, accountsById) => accountId ? accountsById[accountId] : undefined,
+);
+```
+
+---
+
+---
+
+## React Compiler Considerations
+
+**Note:** This codebase uses React Compiler, a build-time tool that automatically optimizes React applications by memoizing components and hooks. React Compiler understands the [Rules of React](https://react.dev/reference/rules) and works with existing JavaScript/TypeScript code without requiring rewrites.
+
+Reference: [React Compiler Introduction](https://github.com/reactwg/react-compiler/discussions/5)
+
+### What React Compiler Does
+
+React Compiler automatically applies memoization to improve update performance (re-renders). It focuses on two main use cases:
+
+1. **Skipping cascading re-rendering of components** - Fine-grained reactivity where only changed parts re-render
+2. **Skipping expensive calculations** - Memoizing expensive computations within components and hooks
+
+#### Example: Automatic Fine-Grained Reactivity
+
+```typescript
+// React Compiler automatically prevents unnecessary re-renders
+function FriendList({ friends }) {
+  const onlineCount = useFriendOnlineCount();
+
+  if (friends.length === 0) {
+    return <NoFriends />;
+  }
+
+  return (
+    <div>
+      <span>{onlineCount} online</span>
+      {friends.map((friend) => (
+        <FriendListCard key={friend.id} friend={friend} />
+      ))}
+      <MessageButton /> {/* Won't re-render when onlineCount changes! */}
+    </div>
+  );
+}
+```
+
+React Compiler determines that `<FriendListCard />` and `<MessageButton />` can be reused even as `friends` or `onlineCount` change, avoiding unnecessary re-renders.
+
+#### Example: Automatic Memoization of Expensive Calculations
+
+```typescript
+// React Compiler automatically memoizes expensive computations
+function TableContainer({ items }) {
+  // This expensive calculation is automatically memoized
+  const data = expensivelyProcessAReallyLargeArrayOfObjects(items);
+  return <Table data={data} />;
+}
+```
+
+**Note:** For truly expensive functions used across multiple components, consider implementing memoization outside React, as React Compiler only memoizes within components/hooks and doesn't share memoization across components.
+
+### React Compiler Assumptions
+
+React Compiler assumes your code:
+
+- ✅ Is valid, semantic JavaScript
+- ✅ Tests nullable/optional values before accessing (e.g., enable `strictNullChecks` in TypeScript)
+- ✅ Follows the [Rules of React](https://react.dev/reference/rules)
+
+React Compiler can verify many Rules of React statically and will **skip compilation** when it detects errors. Install [eslint-plugin-react-compiler](https://www.npmjs.com/package/eslint-plugin-react-compiler) to see compilation errors.
```

Comment:
Why advice here is to install the React compiler eslint plugin, and why don't we add it to the project?